tempesta_fw/http.c

```diff
@@ -223,23 +223,75 @@ tfw_http_prep_304(TfwHttpMsg *resp, TfwHttpReq *req, void *msg_it,
 	if (tfw_http_msg_setup(resp, it, data_len))
 		return TFW_BLOCK;
 
-	tfw_http_msg_write(it, resp, &rh);
+	ret = tfw_http_msg_write(it, resp, &rh);
 	if (end)
-		tfw_http_msg_write(it, resp, end);
+		ret |= tfw_http_msg_write(it, resp, end);
 
 	TFW_DBG("Send HTTP 304 response\n");
 
-	return TFW_PASS;
+	return ret ? TFW_BLOCK : TFW_PASS;
+}
+
+/*
+ * Free an HTTP message.
+ * Also, free the connection instance if there's no more references.
+ *
+ * This function should be used anytime when there's a chance that
+ * a connection instance may belong to multiple messages, which is
+ * almost always. If a connection is suddenly closed then it still
+ * can be safely dereferenced and used in the code.
+ * In rare cases we're sure that a connection instance in a message
+ * doesn't have multiple users. For example, when an error response
+ * is prepared and sent by Tempesta, that HTTP message does not need
+ * a connection instance. The message is then immediately destroyed,
+ * and a simpler tfw_http_msg_free() can be used for that.
+ *
+ * NOTE: @hm->conn may be NULL if @hm is the response that was served
+ * from cache.
+ */
+static void
+tfw_http_conn_msg_free(TfwHttpMsg *hm)
+{
+	if (unlikely(!hm))
+		return;
+
+	if (hm->conn) {
+		/*
+		 * Unlink the connection while there is at least one
+		 * reference. Use atomic exchange to avoid races with
+		 * new messages arriving on the connection.
+		 */
+		__cmpxchg((unsigned long *)&hm->conn->msg, (unsigned long)hm,
+			  0UL, sizeof(long));
+		tfw_connection_put(hm->conn);
+	}
+
+	tfw_http_msg_free(hm);
+}
+
+/*
+ * Close the client connection and free unpaired request. This function
+ * is needed for cases when we cannot prepare response for this request.
+ */
+void
+tfw_http_req_conn_close(TfwHttpReq *req)
+{
+	ss_close_sync(req->conn->sk, true);
+	tfw_http_conn_msg_free((TfwHttpMsg *)req);
+	TFW_INC_STAT_BH(clnt.msgs_otherr);	
```

Comment:
Trailing white space. I noted the issue third time in your patches. Hard to find while reviewing on github.

Reply:
Corrected.